File: csharp/binary-search/BinarySearch.cs
public static class BinarySearch
{
    public static int Find(int[] input, int value)
    {
        if (input.Length == 0)
        {
            return -1;
        }

        var n = input.Length / 2;

        if (value < input[n])
        {
            return Find(input[..n], value);
        } else if (value > input[n])
        {
            var res = Find(input[(n + 1)..], value);

            return res == -1 ? -1 : res + n + 1;
        }

        return n;
    }
}

File: cpp/binary-search/binary_search.cpp
#include "binary_search.h"

#include <stdexcept>

namespace binary_search
{

    size_t find(const std::vector<int> &elements, int x)
    {
        if (elements.size() <= 0)
            throw std::domain_error("Elements must have at least one value");

        size_t left = 0;
        size_t right = elements.size();

        while (left < right)
        {
            size_t n = (right - left) / 2 + left;

            if (x < elements[n])
            {
                right = n;
            }
            else if (x > elements[n])
            {
                left = n + 1;
            }
            else
            {
                return n;
            }
        }

        throw std::domain_error("Value not found");
    }

} // namespace binary_search
